src/System.ServiceProcess.ServiceController/tests/ServiceBaseTests.cs

```diff
@@ -75,104 +76,125 @@ public void Cleanup()
         [ConditionalFact(nameof(IsProcessElevated))]
         public void TestOnStartThenStop()
         {
-            var controller = new ServiceController(_testService.TestServiceName);
-            AssertExpectedProperties(controller);
-            string expected =
-@"OnStart args=
-OnStop
-";
-            controller.Stop();
-            controller.WaitForStatus(ServiceControllerStatus.Stopped);
-            Assert.Equal(expected, _testService.GetServiceOutput());
+            string serviceName = _testService.TestServiceName;
+            using (var client = new NamedPipeClientStream(".", serviceName, PipeDirection.In))
+            {
+                client.Connect();
+                var controller = new ServiceController(serviceName);
+                AssertExpectedProperties(controller);
+
+                controller.Stop();
+                Assert.Equal((int)PipeMessageByteCode.Stop, client.ReadByte());
+            }
         }
 
         [ConditionalFact(nameof(IsProcessElevated))]
         public void TestOnStartWithArgsThenStop()
         {
-            var controller = new ServiceController(_testService.TestServiceName);
-            AssertExpectedProperties(controller);
-            controller.Stop();
-            controller.WaitForStatus(ServiceControllerStatus.Stopped);
-
-            string expected =
-@"OnStart args=a,b,c
-OnStop
-";
-            controller.Start(new string[] { "a", "b", "c" });
-            controller.WaitForStatus(ServiceControllerStatus.Running);
-            controller.Stop();
-            controller.WaitForStatus(ServiceControllerStatus.Stopped);
-            Assert.Equal(expected, _testService.GetServiceOutput());
+            string serviceName = _testService.TestServiceName;
+            var controller = new ServiceController(serviceName);
+
+            using (var client = new NamedPipeClientStream(".", serviceName, PipeDirection.In))
+            {
+                client.Connect();
+                AssertExpectedProperties(controller);
+
+                controller.Stop();
+                Assert.Equal((int)PipeMessageByteCode.Stop, client.ReadByte());
+
+                controller.WaitForStatus(ServiceControllerStatus.Stopped);
+                controller.Start(new string[] { "a", "b", "c" });
+            }
+
+            using (var client = new NamedPipeClientStream(".", serviceName, PipeDirection.In))
```

Comment:
It ought to be possible to keep the previous pipe open rather than need to make another?

Reply:
The client pipe is open and is in connected state. The server is in disconnected state. When the server starts listening again, then this opened pipe is not receiving any messages. I am trying to investigate more why this pipe is not able to receive the messages when the server is live again

Reply:
Hmm, maybe it is not possible to reconnect to a pipe at the server end. It creates logically a new pipe.